numba/core/cgutils.py

```diff
@@ -1013,56 +1014,59 @@ def cbranch_or_continue(builder, cond, bbtrue):
     return bbcont
 
 
-def memcpy(builder, dst, src, count):
+def memcpy(builder, dst, src, count, itemsize=1):
     """
     Emit a memcpy to the builder.
 
-    Copies each element of dst to src. Unlike the C equivalent, each element
-    can be any LLVM type.
+    Source and destination types can be any type, but the types must be the
+    same.
 
-    Assumes
-    -------
-    * dst.type == src.type
-    * count is positive
+    Call either as:
+
+    - memcpy(builder, dst, src, len), or
+    - memcpy(builder, dst, src, count, itemsize)
     """
-    # Note this does seem to be optimized as a raw memcpy() by LLVM
-    # whenever possible...
-    assert dst.type == src.type
-    with for_range(builder, count, intp=count.type) as loop:
-        out_ptr = builder.gep(dst, [loop.index])
-        in_ptr = builder.gep(src, [loop.index])
-        builder.store(builder.load(in_ptr), out_ptr)
+    if dst.type != src.type:
+        msg = f'memcpy requires the same types; got {dst.type} and {src.type}'
+        raise TypingError(msg)
 
+    _raw_mem_intrinsic(builder, 'llvm.memcpy', dst, src, count, itemsize)
 
-def _raw_memcpy(builder, func_name, dst, src, count, itemsize, align):
-    size_t = count.type
-    if isinstance(itemsize, int):
-        itemsize = ir.Constant(size_t, itemsize)
 
-    memcpy = builder.module.declare_intrinsic(func_name,
-                                              [voidptr_t, voidptr_t, size_t])
-    is_volatile = false_bit
-    builder.call(memcpy, [builder.bitcast(dst, voidptr_t),
-                          builder.bitcast(src, voidptr_t),
-                          builder.mul(count, itemsize),
-                          is_volatile])
+def memmove(builder, dst, src, count, itemsize=1):
+    """
+    Emit a memmove to the builder.
 
+    Source and destination types can be any type, but the types must be the
+    same.
```

Comment:
Should this contain a similar check for source type matching dest type as present in the `memcpy` impl above?

Reply:
No, this implementation was OK (although the docstring is wrong, and will be updated).